services/QuillLMS/client/app/bundles/Comprehension/components/studentView/container.tsx

```diff
@@ -106,6 +114,59 @@ export class StudentViewContainer extends React.Component<StudentViewContainerPr
     });
   }
 
+  defaultHandleFinishActivity = () => {
+    this.specifiedActivitySessionUID();
+    // We only post completed sessions if we had one specified when the activity loaded
+    if (!this.specifiedActivitySessionUID()) return
+    const { dispatch, session, } = this.props
+    const { sessionID, } = session
+    const conceptResults = this.generateConceptResults()
+    const percentage = this.calculatePercentage()
+    dispatch(completeActivitySession(sessionID, percentage, conceptResults))
+  }
+
+  calculatePercentage = () => {
+    const { session, } = this.props
+    const { submittedResponses, } = session
+    const attemptCounts = Object.values(submittedResponses).map((responses) => ATTEMPTS_TO_SCORE[responses.length])
+    return attemptCounts.reduce((total, value) => total + value) / attemptCounts.length
+  }
+
+  generateConceptResults = () => {
```

Comment:
I leave this up to your discretion, but I feel like these might be better as lib functions with the arguments passed in than as part of the component. That would make them easier to test, and better isolate them from the rendering logic. That also follows the pattern we have for similar functions in the other apps.

Reply:
Good call.  I've pulled this code into a `libs` directory and added a couple of very basic tests.